### DISCUSSION TITLE
Is Craft Calendar a good fit for work scheduling?

### DISCUSSION THREAD

**Discussion by allesweb** [solspace/craft-calendar#128] [Questions & Answers] [answered]
We're building a work scheduling platform for a client on which users can assign themselves work shifts as volunteers. Shifts can be recurring or during special events. Users accumulate points for working shifts which are counted towards either their own balance or their student association's. Admins can reassign shifts or deny volunteers, which affects their point balance.

Is Craft Calendar a good fit for this use case? Mainly setting up and perhaps planning recurring shifts, allow for user accounts to register these events somehow, count shifts toward a point balance, and also edit these relationships retroactively?

What are some potential challenges and pitfalls if we were to implement Calendar? I guess we're up for a challenge either way, but would Calendar help us on the way? One problem I suspect is that recurring events lack their own ID. Is there any other identifier we can utilize to get around this, such as the date as in the proposed solution? We're of course prepared to do some custom module coding.

**Comment by kjmartens** [comment, marked as answer]
Sorry for the delay @allesweb.

This is a common request we receive. In theory, Calendar is very close to being able to handle hours of operation, shifts, and appointment scheduling. However, it lacks the practicality that most users want to prevent double booking, forcing bookings to be within specific blocks of time and excluding certain days from being booked.

In your particular case, you're also looking for managing/calculating shift hour limits, etc, which is another hurdle that Calendar cannot handle. And as you mention, there is no unique ID for each recurrence. 

While we would be more than happy to earn your business on this one, I do firmly believe this is a very uphill battle to get working for your scenario. 😕

**Reply by allesweb** [reply]
Thanks for your response! 

Since my original post we've installed the Calendar Pro trial, and I actually think we can make this work for us. Perhaps my use case description was too general, but here's my thinking: 

Managers create and manage available shifts in the admin panel, either as a one-off or recurring event. These are the available shifts which are displayed front end for volunteers to assign themselves to through a form. In the case of a recurring event the form creates a proper event mirroring the occurrence. The shift loses the synchronized connection to the original event, but that's ok.

Each event has custom fields for relations, such as the original event (if a mirrored recurring event), staff, reserve staff, etc. We can prevent booking if the staff field is full, we can query shifts and related staff to assign points, we can set the title field to be `{staff.all()|length} assigned to {event.one().title}` for rudimentary meta display, and we can even remove the mirrored occurrence by excluding the date in the original event to mark it as booked.

Currently it seems more like a UX/AX challenge. 

PS. Is there any native way to convert an occurrence to a proper event? That would be neat.

**Reply by kjmartens** [reply]
> PS. Is there any native way to convert an occurrence to a proper event? That would be neat.

This is not currently possible, unfortunately.
